src/briefcase/integrations/android_sdk.py

```diff
@@ -212,61 +248,305 @@ def verify_emulator(self):
                     "platform-tools",
                 ],
                 env=self.env,
-                check=True
+                check=True,
             )
         except subprocess.CalledProcessError:
             raise BriefcaseCommandError(
                 "Error while installing Android emulator and system image."
             )
 
-    def no_or_wrong_device_message(self):
-        adb_path = self.root_path / "platform-tools" / "adb"
-        avdmanager_path = self.root_path / "tools" / "bin" / "avdmanager"
-        emulator_path = self.root_path / "emulator" / "emulator"
-        sdkmanager_path = self.root_path / "tools" / "bin" / "sdkmanager"
-        return """\
-You can get a list of valid devices by running this command:
+    def emulators(self):
+        """Find the list of emulators that are available
 
-    $ {adb_path} devices -l
+        """
+        try:
+            # Capture `stderr` so that if the process exits with failure, the
+            # stderr data is in `e.output`.
+            output = self.command.subprocess.check_output(
+                [str(self.root_path / "emulator" / "emulator"), "-list-avds"],
+                universal_newlines=True,
+                stderr=subprocess.STDOUT,
+            ).strip()
+
+            # AVD names are returned one per line.
+            if len(output) == 0:
+                return []
+            return output.split("\n")
+        except subprocess.CalledProcessError:
+            raise BriefcaseCommandError("Unable to obtain Android emulator list")
+
+    def devices(self):
+        """Find the devices that are attached and available to ADB
+
+        """
+        try:
+            # Capture `stderr` so that if the process exits with failure, the
+            # stderr data is in `e.output`.
+            output = self.command.subprocess.check_output(
+                [str(self.root_path / "platform-tools" / "adb"), "devices", "-l"],
+                universal_newlines=True,
+                stderr=subprocess.STDOUT,
+            ).strip()
+
+            # Process the output of `adb devices -l`.
+            # The first line is header information.
+            # Each subsequent line is a single device descriptor.
+            devices = {}
+            for line in output.split("\n")[1:]:
+                parts = re.sub(r"\s+", " ", line).split(" ")
```

Comment:
I believe that `\t` is a better character to split on for name/rest, but it's OK.

I'm imagining, like:

```
name, property_string = line.split('\t', 1)
properties = property_string.split()
```

Reply:
....except it's not a tab - it's a bunch of spaces.  At least it is in all my testing...